cmd/gateway/s3x/gateway-s3x-ledger.go

```diff
@@ -15,418 +17,84 @@ Any claiming or releasing of locks should be done in the public setter+getter fu
-	}
-	mpart := ledger.MultipartUploads[multipartID]
-	mpart.ObjectParts = append(mpart.ObjectParts, ObjectPartInfo{
-		Number:   partNumber,
-		DataHash: partHash,
-	})
-	ledger.MultipartUploads[multipartID] = mpart
-	return le.putLedger(ledger)
-}
-
-// NewBucket creates a new ledger bucket entry
-func (le *LedgerStore) NewBucket(name, hash string) error {
-	le.Lock()
-	defer le.Unlock()
-	ledger, err := le.getLedger()
-	if err != nil {
-		return err
-	}
-	if le.bucketExists(ledger, name) {
-		return ErrLedgerBucketExists
-	}
-	if ledger.GetBuckets() == nil {
-		ledger.Buckets = make(map[string]LedgerBucketEntry)
-	}
-	ledger.Buckets[name] = LedgerBucketEntry{
-		Objects:  make(map[string]LedgerObjectEntry),
-		Name:     name,
-		IpfsHash: hash,
-	}
-	return le.putLedger(ledger)
-}
-
-// UpdateBucketHash is used to update the ledger bucket entry
-// with a new IPFS hash
-func (le *LedgerStore) UpdateBucketHash(name, hash string) error {
-	le.Lock()
-	defer le.Unlock()
-	ledger, err := le.getLedger()
-	if err != nil {
-		return err
-	}
-	if !le.bucketExists(ledger, name) {
-		return ErrLedgerBucketDoesNotExist
-	}
-	entry := ledger.Buckets[name]
-	entry.IpfsHash = hash
-	ledger.Buckets[name] = entry
-	return le.putLedger(ledger)
-}
-
-// RemoveObject is used to remove a ledger object entry from a ledger bucket entry
-func (le *LedgerStore) RemoveObject(bucketName, objectName string) error {
-	le.Lock()
-	defer le.Unlock()
-	ledger, err := le.getLedger()
-	if err != nil {
-		return err
-	}
-	if err := le.objectExists(ledger, bucketName, objectName); err != nil {
-		return err
-	}
-	delete(ledger.Buckets[bucketName].Objects, objectName)
-	return nil
-}
-
-// AddObjectToBucket is used to update a ledger bucket entry with a new ledger object entry
-func (le *LedgerStore) AddObjectToBucket(bucketName, objectName, objectHash string) error {
-	le.Lock()
-	defer le.Unlock()
-	ledger, err := le.getLedger()
-	if err != nil {
-		return err
-	}
-	if !le.bucketExists(ledger, bucketName) {
-		return ErrLedgerBucketDoesNotExist
-	}
-	// prevent nil map panic
-	if ledger.GetBuckets()[bucketName].Objects == nil {
-		bucket := ledger.Buckets[bucketName]
-		bucket.Objects = make(map[string]LedgerObjectEntry)
-		ledger.Buckets[bucketName] = bucket
-	}
-	ledger.Buckets[bucketName].Objects[objectName] = LedgerObjectEntry{
-		Name:     objectName,
-		IpfsHash: objectHash,
-	}
-	return le.putLedger(ledger)
-}
-
-// DeleteBucket is used to remove a ledger bucket entry
-func (le *LedgerStore) DeleteBucket(name string) error {
-	le.Lock()
-	defer le.Unlock()
-	ledger, err := le.getLedger()
-	if err != nil {
-		return err
-	}
-	if ledger.GetBuckets()[name].Name == "" {
-		return ErrLedgerBucketDoesNotExist
-	}
-	delete(ledger.Buckets, name)
-	return le.putLedger(ledger)
-}
-
 // Close shuts down the ledger datastore
-func (le *LedgerStore) Close() error {
-	le.Lock()
-	defer le.Unlock()
-	return le.ds.Close()
+func (ls *ledgerStore) Close() error {
+	//todo: clean up caches
```

Comment:
shouldn't we lock here to prevent access to the datastore during shutdown?

Reply:
Yes, I have been putting off this part. What if there are other API calls in progress? Also, there should be standard errors when calling the API after close.

Reply:
That's a really good point. I'm honestly not sure what the answer is. I think for now just go with using a lock, and I'll create a GitHub issue to track research as to what the answer is. 

It will probably be something like prevent closing while there are active requests, but prevent new requests from being made